man/dpkg-architecture.1
.\" dpkg manual page - dpkg-architecture(1)
.\"
.\" Copyright © 2005 Marcus Brinkmann <brinkmd@debian.org>
.\" Copyright © 2005 Scott James Remnant <scott@netsplit.com>
.\" Copyright © 2006-2015 Guillem Jover <guillem@debian.org>
.\" Copyright © 2009-2012 Raphaël Hertzog <hertzog@debian.org>
.\"
.\" This is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program.  If not, see <https://www.gnu.org/licenses/>.
.
.TH dpkg\-architecture 1 "2013-03-28" "Debian Project" "dpkg utilities"
.SH NAME
dpkg\-architecture \- set and determine the architecture for package building
.
.SH SYNOPSIS
.B dpkg\-architecture
.RI [ option "...] [" command ]
.PP
.
.SH DESCRIPTION
.B dpkg\-architecture
provides a facility to determine and set the build and
host architecture for package building.
.PP
The build architecture is always determined by an external call to
\fBdpkg\fP(1), and cannot be set at the command line.
.PP
You can specify the host architecture by providing one or both of the options
\fB\-\-host\-arch\fR and \fB\-\-host\-type\fR. The default is determined by
an external call to
.BR gcc (1),
or the same as the build architecture if \fBCC\fP or gcc are both not
available. One out of \fB\-\-host\-arch\fR and \fB\-\-host\-type\fR is
sufficient, the value of the
other will be set to a usable default. Indeed, it is often better to only
specify one, because \fB\%dpkg\-architecture\fP will warn you if your choice
does not match the default.
.
.SH COMMANDS
.TP
.BR \-l ", " \-\-list
Print the environment variables, one each line, in the format
\fIVARIABLE=value\fP. This is the default action.
.TP
.BR \-e ", " \-\-equal " \fIarchitecture\fP"
Check for equality of architecture (since dpkg 1.13.13).
It compares the current or specified Debian host architecture against
\fIarchitecture\fP, to check if they are equal.
This action will not expand the architecture wildcards.
Command finishes with an exit status of 0 if matched, 1 if not matched.

.TP
.BR \-i ", " \-\-is " \fIarchitecture-wildcard\fP"
Check for identity of architecture (since dpkg 1.13.13).
It compares the current or specified Debian host architecture against
\fIarchitecture-wildcard\fP after having expanded it as an architecture
wildcard, to check if they match.
Command finishes with an exit status of 0 if matched, 1 if not matched.
.TP
.BR \-q ", " \-\-query " \fIvariable-name\fP"
Print the value of a single variable.
.TP
.BR \-s ", " \-\-print\-set
Print an export command. This can be used to set the environment variables
using eval.
.TP
.BR \-u ", " \-\-print\-unset
Print a similar command to \fB\-\-print\-unset\fP but to unset all variables.
.TP
.BR \-c ", " \-\-command " \fIcommand\fP"
Execute a \fIcommand\fP in an environment which has all variables set to
the determined value.
.TP
.BR \-L ", " \-\-list\-known
Print a list of valid architecture names.
Possibly restricted by one or more of the matching options
\fB\-\-match\-wildcard\fP, \fB\-\-match\-bits\fP or \fB\-\-match\-endian\fP
(since dpkg 1.17.14).
.TP
.BR \-? ", " \-\-help
Show the usage message and exit.
.TP
.B \-\-version
Show the version and exit.
.
.SH OPTIONS
.TP
.BR \-a ", " \-\-host\-arch " \fIarchitecture\fP"
Set the host Debian architecture.
.TP
.BR \-t ", " \-\-host\-type " \fIgnu-system-type\fP"
Set the host GNU system type.
.TP
.BR \-A ", " \-\-target\-arch " \fIarchitecture\fP"
Set the target Debian architecture (since dpkg 1.17.14).
.TP
.BR \-T ", " \-\-target\-type " \fIgnu-system-type\fP"
Set the target GNU system type (since dpkg 1.17.14).
.TP
.BR \-W ", " \-\-match\-wildcard " \fIarchitecture-wildcard\fP"
Restrict the architectures listed by \fB\-\-list\-known\fP to ones matching
the specified architecture wildcard (since dpkg 1.17.14).
.TP
.BR \-B ", " \-\-match\-bits " \fIarchitecture-bits\fP"
Restrict the architectures listed by \fB\-\-list\-known\fP to ones with the
specified CPU bits (since dpkg 1.17.14). Either \fB32\fP or \fB64\fP.
.TP
.BR \-E ", " \-\-match\-endian " \fIarchitecture-endianness\fP"
Restrict the architectures listed by \fB\-\-list\-known\fP to ones with the
specified endianness (since dpkg 1.17.14). Either \fBlittle\fP or \fBbig\fP.
.TP
.BR \-f ", " \-\-force
Values set by existing environment variables with the same name as used by
the scripts are honored (i.e. used by \fB\%dpkg\-architecture\fP), except if
this force flag is present. This allows the user
to override a value even when the call to \fB\%dpkg\-architecture\fP is buried
in some other script (for example \fB\%dpkg\-buildpackage\fP(1)).
.
.SH TERMS
.IP "build machine" 4
The machine the package is built on.
.IP "host machine" 4
The machine the package is built for.
.IP "target machine" 4
The machine the compiler is building for.
This is only needed when building a cross-toolchain, one that will be built
on the build architecture, to be run on the host architecture, and to build
code for the target architecture.
.IP "Debian architecture" 4
The Debian architecture string, which specifies the binary tree in the
FTP archive. Examples: i386, sparc, hurd\-i386.
.IP "Debian architecture wildcard" 4
A Debian architecture wildcard is a special architecture string that will match
any real architecture being part of it. The general form is <kernel>\-<cpu>.
Examples: linux\-any, any\-i386, hurd\-any.
.IP "GNU system type" 4
An architecture specification string consisting of two parts separated by
a hyphen: cpu and system.
Examples: i586\-linux\-gnu, sparc\-linux\-gnu, i686\-gnu, x86_64\-netbsd.
.IP "multiarch triplet" 4
The clarified GNU system type, used for filesystem paths.
This triplet does not change even when the baseline ISA gets bumped,
so that the resulting paths are stable over time.
The only current difference with the GNU system type is that the CPU part
for i386 based systems is always i386.
Examples: i386\-linux\-gnu, x86_64\-linux\-gnu.
Example paths: /lib/powerpc64le-linux-gnu/, /usr/lib/i386-kfreebsd-gnu/.
.
.SH VARIABLES
The following variables are set by \fBdpkg\-architecture\fP:
.IP "\fBDEB_BUILD_ARCH\fP" 4
The Debian architecture of the build machine.
.IP "\fBDEB_BUILD_ARCH_OS\fP" 4
The Debian system name of the build machine (since dpkg 1.13.2).
.IP "\fBDEB_BUILD_ARCH_CPU\fP" 4
The Debian cpu name of the build machine (since dpkg 1.13.2).
.IP "\fBDEB_BUILD_ARCH_BITS\fP" 4
The pointer size of the build machine (in bits; since dpkg 1.15.4).
.IP "\fBDEB_BUILD_ARCH_ENDIAN\fP" 4
The endianness of the build machine (little / big; since dpkg 1.15.4).
.IP "\fBDEB_BUILD_GNU_CPU\fP" 4
The CPU part of \fBDEB_BUILD_GNU_TYPE\fP.
.IP "\fBDEB_BUILD_GNU_SYSTEM\fP" 4
The System part of \fBDEB_BUILD_GNU_TYPE\fP.
.IP "\fBDEB_BUILD_GNU_TYPE\fP" 4
The GNU system type of the build machine.
.IP "\fBDEB_BUILD_MULTIARCH\fP" 4
The clarified GNU system type of the build machine, used for filesystem
paths (since dpkg 1.16.0).
.IP "\fBDEB_HOST_ARCH\fP" 4
The Debian architecture of the host machine.
.IP "\fBDEB_HOST_ARCH_OS\fP" 4
The Debian system name of the host machine (since dpkg 1.13.2).
.IP "\fBDEB_HOST_ARCH_CPU\fP" 4
The Debian cpu name of the host machine (since dpkg 1.13.2).
.IP "\fBDEB_HOST_ARCH_BITS\fP" 4
The pointer size of the host machine (in bits; since dpkg 1.15.4).
.IP "\fBDEB_HOST_ARCH_ENDIAN\fP" 4
The endianness of the host machine (little / big; since dpkg 1.15.4).
.IP "\fBDEB_HOST_GNU_CPU\fP" 4
The CPU part of \fBDEB_HOST_GNU_TYPE\fP.
.IP "\fBDEB_HOST_GNU_SYSTEM\fP" 4
The System part of \fBDEB_HOST_GNU_TYPE\fP.
.IP "\fBDEB_HOST_GNU_TYPE\fP" 4
The GNU system type of the host machine.
.IP "\fBDEB_HOST_MULTIARCH\fP" 4
The clarified GNU system type of the host machine, used for filesystem
paths (since dpkg 1.16.0).
.IP "\fBDEB_TARGET_ARCH\fP" 4
The Debian architecture of the target machine (since dpkg 1.17.14).
.IP "\fBDEB_TARGET_ARCH_OS\fP" 4
The Debian system name of the target machine (since dpkg 1.17.14).
.IP "\fBDEB_TARGET_ARCH_CPU\fP" 4
The Debian cpu name of the target machine (since dpkg 1.17.14).
.IP "\fBDEB_TARGET_ARCH_BITS\fP" 4
The pointer size of the target machine (in bits; since dpkg 1.17.14).
.IP "\fBDEB_TARGET_ARCH_ENDIAN\fP" 4
The endianness of the target machine (little / big; since dpkg 1.17.14).
.IP "\fBDEB_TARGET_GNU_CPU\fP" 4
The CPU part of \fBDEB_TARGET_GNU_TYPE\fP (since dpkg 1.17.14).
.IP "\fBDEB_TARGET_GNU_SYSTEM\fP" 4
The System part of \fBDEB_TARGET_GNU_TYPE\fP (since dpkg 1.17.14).
.IP "\fBDEB_TARGET_GNU_TYPE\fP" 4
The GNU system type of the target machine (since dpkg 1.17.14).
.IP "\fBDEB_TARGET_MULTIARCH\fP" 4
The clarified GNU system type of the target machine, used for filesystem
paths (since dpkg 1.17.14).
.
.SH FILES
.SS Architecture tables
All these files have to be present for \fBdpkg\-architecture\fP to
work. Their location can be overridden at runtime with the environment
variable \fBDPKG_DATADIR\fP.
.TP
.I /usr/share/dpkg/cputable
Table of known CPU names and mapping to their GNU name.
.TP
.I /usr/share/dpkg/ostable
Table of known operating system names and mapping to their GNU name.
.TP
.I /usr/share/dpkg/triplettable
Mapping between Debian architecture triplets and Debian architecture
names.
.TP
.I /usr/share/dpkg/abitable
Table of Debian architecture ABI attribute overrides.
.SS Packaging support
.TP
.I /usr/share/dpkg/architecture.mk
Makefile snippet that properly sets and exports all the variables that
\fBdpkg\-architecture\fP outputs (since dpkg 1.16.1).
.
.SH EXAMPLES
\fBdpkg\-buildpackage\fP accepts the \fB\-a\fR option and passes it to
\fB\%dpkg\-architecture\fP. Other examples:
.IP
CC=i386\-gnu\-gcc dpkg\-architecture \-c debian/rules build
.IP
eval \`dpkg\-architecture \-u\`
.PP
Check if the current or specified host architecture is equal to an
architecture:
.IP
dpkg\-architecture \-elinux\-alpha
.IP
dpkg\-architecture \-amips \-elinux\-mips
.PP
Check if the current or specified host architecture is a Linux system:
.IP
dpkg\-architecture \-ilinux\-any
.IP
dpkg\-architecture \-ai386 \-ilinux\-any
.
.SS Usage in debian/rules
The environment variables set by \fBdpkg\-architecture\fP are passed to
\fIdebian/rules\fP as make variables (see make documentation). However,
you should not rely on them, as this breaks manual invocation of the
script. Instead, you should always initialize them using
\fB\%dpkg\-architecture\fP with the \fB\-q\fP option. Here are some examples,
which also show how you can improve the cross compilation support in your
package:
.PP
Retrieving the GNU system type and forwarding it to ./configure:
.PP
.RS 4
.nf
DEB_BUILD_GNU_TYPE ?= $(shell dpkg\-architecture \-qDEB_BUILD_GNU_TYPE)
DEB_HOST_GNU_TYPE ?= $(shell dpkg\-architecture \-qDEB_HOST_GNU_TYPE)
[...]
ifeq ($(DEB_BUILD_GNU_TYPE), $(DEB_HOST_GNU_TYPE))
  confflags += \-\-build=$(DEB_HOST_GNU_TYPE)
else
  confflags += \-\-build=$(DEB_BUILD_GNU_TYPE) \\
               \-\-host=$(DEB_HOST_GNU_TYPE)
endif
[...]
\&./configure $(confflags)
.fi
.RE
.PP
Doing something only for a specific architecture:
.PP
.RS 4
.nf
DEB_HOST_ARCH ?= $(shell dpkg\-architecture \-qDEB_HOST_ARCH)

ifeq ($(DEB_HOST_ARCH),alpha)
  [...]
endif
.fi
.RE
.PP
or if you only need to check the CPU or OS type, use the
\fBDEB_HOST_ARCH_CPU\fP or \fBDEB_HOST_ARCH_OS\fP variables.
.PP
Note that you can also rely on an external Makefile snippet to properly
set all the variables that \fBdpkg\-architecture\fP can provide:
.PP
.RS 4
.nf
include /usr/share/dpkg/architecture.mk

ifeq ($(DEB_HOST_ARCH),alpha)
  [...]
endif
.fi
.RE
.PP
In any case, you should never use \fBdpkg \-\-print\-architecture\fP to get
architecture information during a package build.
.
.SH NOTES
All long command and option names available only since dpkg 1.17.17.
.
.SH SEE ALSO
.BR dpkg\-buildpackage (1),
.BR dpkg\-cross (1).


scripts/dpkg-architecture.pl
#!/usr/bin/perl
#
# dpkg-architecture
#
# Copyright © 1999-2001 Marcus Brinkmann <brinkmd@debian.org>
# Copyright © 2004-2005 Scott James Remnant <scott@netsplit.com>,
# Copyright © 2006-2014 Guillem Jover <guillem@debian.org>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

use strict;
use warnings;

use Dpkg ();
use Dpkg::Gettext;
use Dpkg::Getopt;
use Dpkg::ErrorHandling;
use Dpkg::Arch qw(get_raw_build_arch get_raw_host_arch get_host_gnu_type
                  debarch_to_cpuattrs
                  get_valid_arches debarch_eq debarch_is debarch_to_debtuple
                  debarch_to_gnutriplet gnutriplet_to_debarch
                  debarch_to_multiarch);

textdomain('dpkg-dev');

sub version {
    printf g_("Debian %s version %s.\n"), $Dpkg::PROGNAME, $Dpkg::PROGVERSION;

    printf g_('
This is free software; see the GNU General Public License version 2 or
later for copying conditions. There is NO warranty.
');
}

sub usage {
    printf g_(
'Usage: %s [<option>...] [<command>]')
    . "\n\n" . g_(
'Commands:
  -l, --list                list variables (default).
  -L, --list-known          list valid architectures (matching some criteria).
  -e, --equal <arch>        compare with host Debian architecture.
  -i, --is <arch-wildcard>  match against host Debian architecture.
  -q, --query <variable>    prints only the value of <variable>.
  -s, --print-set           print command to set environment variables.
  -u, --print-unset         print command to unset environment variables.
  -c, --command <command>   set environment and run the command in it.
  -?, --help                show this help message.
      --version             show the version.')
    . "\n\n" . g_(
'Options:
  -a, --host-arch <arch>    set host Debian architecture.
  -t, --host-type <type>    set host GNU system type.
  -A, --target-arch <arch>  set target Debian architecture.
  -T, --target-type <type>  set target GNU system type.
  -W, --match-wildcard <arch-wildcard>
                            restrict architecture list matching <arch-wildcard>.
  -B, --match-bits <arch-bits>
                            restrict architecture list matching <arch-bits>.
  -E, --match-endian <arch-endian>
                            restrict architecture list matching <arch-endian>.
  -f, --force               force flag (override variables set in environment).')
    . "\n", $Dpkg::PROGNAME;
}

sub check_arch_coherency
{
    my ($arch, $gnu_type) = @_;

    if ($arch ne '' && $gnu_type eq '') {
        $gnu_type = debarch_to_gnutriplet($arch);
        error(g_('unknown Debian architecture %s, you must specify ' .
                 'GNU system type, too'), $arch)
            unless defined $gnu_type;
    }

    if ($gnu_type ne '' && $arch eq '') {
        $arch = gnutriplet_to_debarch($gnu_type);
        error(g_('unknown GNU system type %s, you must specify ' .
                 'Debian architecture, too'), $gnu_type)
            unless defined $arch;
    }

    if ($gnu_type ne '' && $arch ne '') {
        my $dfl_gnu_type = debarch_to_gnutriplet($arch);
        error(g_('unknown default GNU system type for Debian architecture %s'),
              $arch)
            unless defined $dfl_gnu_type;
        warning(g_('default GNU system type %s for Debian arch %s does not ' .
                   'match specified GNU system type %s'), $dfl_gnu_type,
                $arch, $gnu_type)
            if $dfl_gnu_type ne $gnu_type;
    }

    return ($arch, $gnu_type);
}

use constant {
    DEB_NONE => 0,
    DEB_BUILD => 1,
    DEB_HOST => 2,
    DEB_TARGET => 64,
    DEB_ARCH_INFO => 4,
    DEB_ARCH_ATTR => 8,
    DEB_MULTIARCH => 16,
    DEB_GNU_INFO => 32,
};

use constant DEB_ALL => DEB_BUILD | DEB_HOST | DEB_TARGET |
                        DEB_ARCH_INFO | DEB_ARCH_ATTR |
                        DEB_MULTIARCH | DEB_GNU_INFO;

my %arch_vars = (
    DEB_BUILD_ARCH => DEB_BUILD,
    DEB_BUILD_ARCH_ABI => DEB_BUILD | DEB_ARCH_INFO,
    DEB_BUILD_ARCH_LIBC => DEB_BUILD | DEB_ARCH_INFO,
    DEB_BUILD_ARCH_OS => DEB_BUILD | DEB_ARCH_INFO,
    DEB_BUILD_ARCH_CPU => DEB_BUILD | DEB_ARCH_INFO,
    DEB_BUILD_ARCH_BITS => DEB_BUILD | DEB_ARCH_ATTR,
    DEB_BUILD_ARCH_ENDIAN => DEB_BUILD | DEB_ARCH_ATTR,
    DEB_BUILD_MULTIARCH => DEB_BUILD | DEB_MULTIARCH,
    DEB_BUILD_GNU_CPU => DEB_BUILD | DEB_GNU_INFO,
    DEB_BUILD_GNU_SYSTEM => DEB_BUILD | DEB_GNU_INFO,
    DEB_BUILD_GNU_TYPE => DEB_BUILD | DEB_GNU_INFO,
    DEB_HOST_ARCH => DEB_HOST,
    DEB_HOST_ARCH_ABI => DEB_HOST | DEB_ARCH_INFO,
    DEB_HOST_ARCH_LIBC => DEB_HOST | DEB_ARCH_INFO,
    DEB_HOST_ARCH_OS => DEB_HOST | DEB_ARCH_INFO,
    DEB_HOST_ARCH_CPU => DEB_HOST | DEB_ARCH_INFO,
    DEB_HOST_ARCH_BITS => DEB_HOST | DEB_ARCH_ATTR,
    DEB_HOST_ARCH_ENDIAN => DEB_HOST | DEB_ARCH_ATTR,
    DEB_HOST_MULTIARCH => DEB_HOST | DEB_MULTIARCH,
    DEB_HOST_GNU_CPU => DEB_HOST | DEB_GNU_INFO,
    DEB_HOST_GNU_SYSTEM => DEB_HOST | DEB_GNU_INFO,
    DEB_HOST_GNU_TYPE => DEB_HOST | DEB_GNU_INFO,
    DEB_TARGET_ARCH => DEB_TARGET,
    DEB_TARGET_ARCH_ABI => DEB_TARGET | DEB_ARCH_INFO,
    DEB_TARGET_ARCH_LIBC => DEB_TARGET | DEB_ARCH_INFO,
    DEB_TARGET_ARCH_OS => DEB_TARGET | DEB_ARCH_INFO,
    DEB_TARGET_ARCH_CPU => DEB_TARGET | DEB_ARCH_INFO,
    DEB_TARGET_ARCH_BITS => DEB_TARGET | DEB_ARCH_ATTR,
    DEB_TARGET_ARCH_ENDIAN => DEB_TARGET | DEB_ARCH_ATTR,
    DEB_TARGET_MULTIARCH => DEB_TARGET | DEB_MULTIARCH,
    DEB_TARGET_GNU_CPU => DEB_TARGET | DEB_GNU_INFO,
    DEB_TARGET_GNU_SYSTEM => DEB_TARGET | DEB_GNU_INFO,
    DEB_TARGET_GNU_TYPE => DEB_TARGET | DEB_GNU_INFO,
);

my $req_vars = DEB_ALL;
my $req_host_arch = '';
my $req_host_gnu_type = '';
my $req_target_arch = '';
my $req_target_gnu_type = '';
my $req_eq_arch = '';
my $req_is_arch = '';
my $req_match_wildcard = '';
my $req_match_bits = '';
my $req_match_endian = '';
my $req_variable_to_print;
my $action = 'list';
my $force = 0;

sub action_needs($) {
  my $bits = shift;
  return (($req_vars & $bits) == $bits);
}

@ARGV = normalize_options(@ARGV);

while (@ARGV) {
    my $arg = shift;

    if ($arg eq '-a' or $arg eq '--host-arch') {
	$req_host_arch = shift;
    } elsif ($arg eq '-t' or $arg eq '--host-type') {
	$req_host_gnu_type = shift;
    } elsif ($arg eq '-A' or $arg eq '--target-arch') {
	$req_target_arch = shift;
    } elsif ($arg eq '-T' or $arg eq '--target-type') {
	$req_target_gnu_type = shift;
    } elsif ($arg eq '-W' or $arg eq '--match-wildcard') {
	$req_match_wildcard = shift;
    } elsif ($arg eq '-B' or $arg eq '--match-bits') {
	$req_match_bits = shift;
    } elsif ($arg eq '-E' or $arg eq '--match-endian') {
	$req_match_endian = shift;
    } elsif ($arg eq '-e' or $arg eq '--equal') {
	$req_eq_arch = shift;
	$req_vars = $arch_vars{DEB_HOST_ARCH};
	$action = 'equal';
    } elsif ($arg eq '-i' or $arg eq '--is') {
	$req_is_arch = shift;
	$req_vars = $arch_vars{DEB_HOST_ARCH};
	$action = 'is';
    } elsif ($arg eq '-u' or $arg eq '--print-unset') {
	$req_vars = DEB_NONE;
	$action = 'print-unset';
    } elsif ($arg eq '-l' or $arg eq '--list') {
	$action = 'list';
    } elsif ($arg eq '-s' or $arg eq '--print-set') {
	$req_vars = DEB_ALL;
	$action = 'print-set';
    } elsif ($arg eq '-f' or $arg eq '--force') {
        $force=1;
    } elsif ($arg eq '-q' or $arg eq '--query') {
	my $varname = shift;
	error(g_('%s is not a supported variable name'), $varname)
	    unless (exists $arch_vars{$varname});
	$req_variable_to_print = "$varname";
	$req_vars = $arch_vars{$varname};
        $action = 'query';
    } elsif ($arg eq '-c' or $arg eq '--command') {
       $action = 'command';
       last;
    } elsif ($arg eq '-L' or $arg eq '--list-known') {
        $req_vars = 0;
        $action = 'list-known';
    } elsif ($arg eq '-?' or $arg eq '--help') {
        usage();
       exit 0;
    } elsif ($arg eq '--version') {
        version();
       exit 0;
    } else {
        usageerr(g_("unknown option '%s'"), $arg);
    }
}

my %v;

#
# Set build variables
#

$v{DEB_BUILD_ARCH} = get_raw_build_arch()
    if (action_needs(DEB_BUILD));
($v{DEB_BUILD_ARCH_ABI}, $v{DEB_BUILD_ARCH_LIBC},
 $v{DEB_BUILD_ARCH_OS}, $v{DEB_BUILD_ARCH_CPU}) = debarch_to_debtuple($v{DEB_BUILD_ARCH})
    if (action_needs(DEB_BUILD | DEB_ARCH_INFO));
($v{DEB_BUILD_ARCH_BITS}, $v{DEB_BUILD_ARCH_ENDIAN}) = debarch_to_cpuattrs($v{DEB_BUILD_ARCH})
    if (action_needs(DEB_BUILD | DEB_ARCH_ATTR));

$v{DEB_BUILD_MULTIARCH} = debarch_to_multiarch($v{DEB_BUILD_ARCH})
    if (action_needs(DEB_BUILD | DEB_MULTIARCH));

if (action_needs(DEB_BUILD | DEB_GNU_INFO)) {
  $v{DEB_BUILD_GNU_TYPE} = debarch_to_gnutriplet($v{DEB_BUILD_ARCH});
  ($v{DEB_BUILD_GNU_CPU}, $v{DEB_BUILD_GNU_SYSTEM}) = split(/-/, $v{DEB_BUILD_GNU_TYPE}, 2);
}

#
# Set host variables
#

# First perform some sanity checks on the host arguments passed.

($req_host_arch, $req_host_gnu_type) = check_arch_coherency($req_host_arch, $req_host_gnu_type);

# Proceed to compute the host variables if needed.

$v{DEB_HOST_ARCH} = $req_host_arch || get_raw_host_arch()
    if (action_needs(DEB_HOST));
($v{DEB_HOST_ARCH_ABI}, $v{DEB_HOST_ARCH_LIBC},
 $v{DEB_HOST_ARCH_OS}, $v{DEB_HOST_ARCH_CPU}) = debarch_to_debtuple($v{DEB_HOST_ARCH})
    if (action_needs(DEB_HOST | DEB_ARCH_INFO));
($v{DEB_HOST_ARCH_BITS}, $v{DEB_HOST_ARCH_ENDIAN}) = debarch_to_cpuattrs($v{DEB_HOST_ARCH})
    if (action_needs(DEB_HOST | DEB_ARCH_ATTR));

$v{DEB_HOST_MULTIARCH} = debarch_to_multiarch($v{DEB_HOST_ARCH})
    if (action_needs(DEB_HOST | DEB_MULTIARCH));

if (action_needs(DEB_HOST | DEB_GNU_INFO)) {
    if ($req_host_gnu_type eq '') {
        $v{DEB_HOST_GNU_TYPE} = debarch_to_gnutriplet($v{DEB_HOST_ARCH});
    } else {
        $v{DEB_HOST_GNU_TYPE} = $req_host_gnu_type;
    }
    ($v{DEB_HOST_GNU_CPU}, $v{DEB_HOST_GNU_SYSTEM}) = split(/-/, $v{DEB_HOST_GNU_TYPE}, 2);

    my $host_gnu_type = get_host_gnu_type();

    warning(g_('specified GNU system type %s does not match CC system ' .
               'type %s, try setting a correct CC environment variable'),
            $v{DEB_HOST_GNU_TYPE}, $host_gnu_type)
        if ($host_gnu_type ne '') && ($host_gnu_type ne $v{DEB_HOST_GNU_TYPE});
}

#
# Set target variables
#

# First perform some sanity checks on the target arguments passed.

($req_target_arch, $req_target_gnu_type) = check_arch_coherency($req_target_arch, $req_target_gnu_type);

# Proceed to compute the target variables if needed.

$v{DEB_TARGET_ARCH} = $req_target_arch || $req_host_arch || get_raw_host_arch()
    if (action_needs(DEB_TARGET));
($v{DEB_TARGET_ARCH_ABI}, $v{DEB_TARGET_ARCH_LIBC},
 $v{DEB_TARGET_ARCH_OS}, $v{DEB_TARGET_ARCH_CPU}) = debarch_to_debtuple($v{DEB_TARGET_ARCH})
    if (action_needs(DEB_TARGET | DEB_ARCH_INFO));
($v{DEB_TARGET_ARCH_BITS}, $v{DEB_TARGET_ARCH_ENDIAN}) = debarch_to_cpuattrs($v{DEB_TARGET_ARCH})
    if (action_needs(DEB_TARGET | DEB_ARCH_ATTR));

$v{DEB_TARGET_MULTIARCH} = debarch_to_multiarch($v{DEB_TARGET_ARCH})
    if (action_needs(DEB_TARGET | DEB_MULTIARCH));

if (action_needs(DEB_TARGET | DEB_GNU_INFO)) {
    if ($req_target_gnu_type eq '') {
        $v{DEB_TARGET_GNU_TYPE} = debarch_to_gnutriplet($v{DEB_TARGET_ARCH});
    } else {
        $v{DEB_TARGET_GNU_TYPE} = $req_target_gnu_type;
    }
    ($v{DEB_TARGET_GNU_CPU}, $v{DEB_TARGET_GNU_SYSTEM}) = split(/-/, $v{DEB_TARGET_GNU_TYPE}, 2);
}


for my $k (keys %arch_vars) {
    $v{$k} = $ENV{$k} if (length $ENV{$k} && !$force);
}

if ($action eq 'list') {
    foreach my $k (sort keys %arch_vars) {
	print "$k=$v{$k}\n";
    }
} elsif ($action eq 'print-set') {
    foreach my $k (sort keys %arch_vars) {
	print "$k=$v{$k}; ";
    }
    print 'export ' . join(' ', sort keys %arch_vars) . "\n";
} elsif ($action eq 'print-unset') {
    print 'unset ' . join(' ', sort keys %arch_vars) . "\n";
} elsif ($action eq 'equal') {
    exit !debarch_eq($v{DEB_HOST_ARCH}, $req_eq_arch);
} elsif ($action eq 'is') {
    exit !debarch_is($v{DEB_HOST_ARCH}, $req_is_arch);
} elsif ($action eq 'command') {
    @ENV{keys %v} = values %v;
    exec @ARGV;
} elsif ($action eq 'query') {
    print "$v{$req_variable_to_print}\n";
} elsif ($action eq 'list-known') {
    foreach my $arch (get_valid_arches()) {
        my ($bits, $endian) = debarch_to_cpuattrs($arch);

        next if $req_match_endian and $endian ne $req_match_endian;
        next if $req_match_bits and $bits ne $req_match_bits;
        next if $req_match_wildcard and not debarch_is($arch, $req_match_wildcard);

        print "$arch\n";
    }
}
